### DISCUSSION TITLE
音樂背景播放與音樂裁切教學

### DISCUSSION THREAD

**Discussion by ghost** [1101-GS4719-C/Python-Community-GS4719-C#371] [General]
在寫code時，偶爾會想讓自己的音樂播放出來，可能做為測試抑或是某個效果
方法有很多，這邊舉例三種(不含 pygame )

- 使用默認撥放器撥放(幾乎適用能在電腦播放的音檔，甚至包含midi,ogg..等等)
```
import os
os.chdir("./music")
os.system("River_Flows_In_You.m4a")
```
第二行是因為我有在當前目錄建立一資料夾存音檔，故須將當前目錄跳至此music資料夾，若音檔就在當前目錄則不用
**system函數可以將字串轉化成命令在伺服器上運行；其原理是每一條system函數執行時，其會創建一個子進程在系統上執行命令行，
要注意，子進程的執行結果無法影響主進程(**註一**)**
執行後效果如下：

![image](https://user-images.githubusercontent.com/100221228/155874135-79877256-2cef-4c82-a8b5-0a5b9c4d2962.png)

不同的檔，開啟方式可能不同，比如midi檔則是使用windows media player開啟；ogg檔可能會用到外在的撥放器，比如Pot Player

- 背景播放，不支援m4a，多以主流的mp3與wav檔為主
wav較需要注意，若出現下方error，需添加"./"於前方
![image](https://user-images.githubusercontent.com/100221228/155874889-d6a15126-e1fe-435a-a207-9720fb0e5958.png)
正確寫法如下
```
from playsound import playsound
import os
os.chdir("./music")
playsound("./River_Flows_In_You.wav")
```
或是將裡面直接貼上完整的路徑資料也行，此為.wav的背景撥放

mp3就較不會有上方問題，但一般添加./能更保險
```
from playsound import playsound
import os
os.chdir("./music")
playsound("cry.mp3")
```

- 背景播放(與上一個類似，且兼容m4a)
不囉嗦，程式如下：
```
import os
from pydub import *
from pydub.playback import play
os.chdir("./music")
play(AudioSegment.from_file("River_Flows_In_You.m4a"))
```
pydub 是個人，比較飯用的，因為他同時兼具了load音樂的功能
`A = AudioSegment.from_file("River_Flows_In_You.m4a")`
其實就是代表了A load此音檔，並可額外做其他事情，比如轉成mp3，以wav檔裁切

這裡順便給個專門裁切的函式：
```
from pydub import *
import os
def cut_music(Object, start, finish):   # 注意：object不能含檔名
    path = os.getcwd()
    os.chdir("./music")
    # 讀取同字串之音檔名之wav檔(object)
    input_music = AudioSegment.from_wav(Object+".wav")
    # 輸出你給進去從幾毫秒開始到幾毫秒結束的裁切音檔
    output_music = input_music[start:finish]
    # 因為單位是毫秒，所以除以1000得秒數
    second = (finish - start)/1000
    # 將此裁切音檔輸出並以wav檔存檔，檔名用str()將剛剛的秒數嵌入，以利辨識
    output_music.export(Object+"_"+str(second)+"_cut_output.wav", format="wav")
    os.chdir(path)  # 回到原始目錄

if __name__ == '__main__':
    # 測試，我想要裁切10s~12s的段落
    cut_music('River_Flows_In_You', start=10000, finish=12000)
```
pydub可做基本的音檔處理，也適用wav,m4a,mp3等常用音檔
若有興趣了解可查詢相關音檔處理的教學，更深入的音檔辨識，可能會使用到libosa之類的相關套件，
這邊顯粗略地介紹一些撥放音檔的方法，與延伸的裁切code
不過有些場合可能不適用，比如使用pygame製作遊戲，通常是使用內建mixer函數實現。

謝謝~

註一：https://www.cnblogs.com/cwp-bg/p/8465566.html

**Comment by handsomeniko** [comment]
蠻猛的ㄟ，先記起來~~

**Comment by pekoranoinu** [comment]
原來python還有這種用途，這一定要學起來。

**Comment by chupochun** [comment]
好帥 居然可以設計程式讓音樂自動撥放
